backend/agentd/agentd.go

```diff
@@ -31,22 +33,36 @@ type Store interface {
 
 // Agentd is the backend HTTP API.
 type Agentd struct {
+	// Host is the hostname Agentd is running on.
+	Host string
+
+	// Port is the port Agentd is running on.
+	Port int
+
 	stopping   chan struct{}
 	running    *atomic.Value
 	wg         *sync.WaitGroup
 	errChan    chan error
 	httpServer *http.Server
+	store      Store
+	bus        messaging.MessageBus
+	tls        *types.TLSOptions
+}
 
-	Store      Store
-	Host       string
-	Port       int
-	MessageBus messaging.MessageBus
-	TLS        *types.TLSOptions
+// New creates a new Agentd.
+func New(client *clientv3.Client, etcdName, host string, port int, bus messaging.MessageBus, tls *types.TLSOptions) *Agentd {
```

Comment:
I don't necessarily have all the context behind https://github.com/sensu/sensu-go/issues/987 but I'm a bit concerned about exposing etcd directly within the daemons.

I loved the idea of having a _generic_ store interface which hides all implementation details of the actual datastore. But maybe that's not relevant anymore (e.g. we do not plan on supporting an another store) or there's more cons than pros to that approach!

Reply:
@palourde,

The daemons themselves still use a generic interface for storage, we just create that storage within the `New` function. How would you feel about a `NewWithEtcdStore` name instead?

Reply:
@echlebek Sorry, I wasn't really clear. I meant exposing the etcd package into the daemon packages.

Currently in master, if I wanted to introduce a new datastore other than etcd, I could simply replace the store that get created & passed in `backend/backend.go`. With this PR, the implementation details of the datastore get propagated within the daemon packages. But maybe it's not a bad thing, I'm just concerned it removes an abstraction layer of the store.

Am I making sense?

Reply:
Let’s be clear. One does not simply replace etcd. ;)

Reply:
Alright, let's ignore my comment then :)

Reply:
@palourde @grepory No, I think Simon's comment actually makes sense. Here is what the Newers would look like if we didn't pass in the client:

```
func New(store store.Store, qg types.QueueGetter, rg types.RingGetter, bus messaging.MessageBus, host string, port int) * Agentd {
}
```

My thought was that we could write new Newers if we ever replaced etcd. I agree with Greg that it would be a non-trivial effort.

We could also do something like this:

```
type DaemonConfig struct {
    Store store.Store
    QueueGetter types.QueueGetter
    RingGetter types.RingGetter
    // blah blah blah
}
```

Then we could add fields without breaking backwards compatibility, and pass a single field in to the Newers.

Reply:
I like this idea of `DaemonConfig`!